### ISSUE TITLE
Make mountains snowy?

### ISSUE THREAD

**Issue by cooljeanius** [cooljeanius/Son_Of_The_Black_Eye_Easy_Edition#10]
In the winter levels, the mountains are just regular mountains (`Mm`) instead of snow-covered ones (`Ms`), which has the odd effect of having some units be able to move more quickly thru the mountains than they do across the plains. @nemaara thinks that this is because the original SotBE might have been written before snow-covered mountains were introduced, but I would want to verify that before changing anything here. Also, making the mountains snow-covered might make things harder, when I'm trying to make things easier in this campaign, so I'd want to consider that, as well... I might still make the change anyways, though, just because it'd make more sense...

**Comment by cooljeanius** [comment]
Hey @Ashmyr and @knyghtmare what are your opinions on whether I should do this?

**Comment by brirush84** [comment]
I definitely used the mountains as a shortcut around the plains in the snow scenarios. Slowing everything down would be super painful, but snowy mountains make more sense.

I need to try the easier version (since I saw on another issue that you added ice monsters), but have you thought of having a couple of 'snowshoe' items that players find in the first snow level that some units can equip that improves snow speed? Just having 2-3 powerful/favourite units that can go fast would help a lot.

**Comment by cooljeanius** [comment]
> I definitely used the mountains as a shortcut around the plains in the snow scenarios. Slowing everything down would be super painful, but snowy mountains make more sense.
> 
> I need to try the easier version (since I saw on another issue that you added ice monsters), but have you thought of having a couple of 'snowshoe' items that players find in the first snow level that some units can equip that improves snow speed? Just having 2-3 powerful/favourite units that can go fast would help a lot.

yeah, that was something I was considering doing with bug #6; I just haven't managed to figure out how to do so yet; creating new unique pickuppable items isn't something I've done in general yet... maybe I can see if some other add-on already has some snowshoes I can steal...

**Comment by brirush84** [comment]
This code for making movement impossible might help:

```
#define MOVEMENT_RESTRICTION
    # to avoid luring onto unprotected terrain, can still move in castles
    [object]
        silent=yes
        duration=scenario
        [effect]
            apply_to=movement_costs
            replace=true
            [movement_costs]
                flat={UNREACHABLE}
                frozen={UNREACHABLE}
                cave={UNREACHABLE}
                reef={UNREACHABLE}
                sand={UNREACHABLE}
                shallow_water={UNREACHABLE}
                swamp_water={UNREACHABLE}
                deep_water={UNREACHABLE}
            [/movement_costs]
        [/effect]
    [/object]
#enddef
```

We could just get rid of the others and make 'frozen=1' or something, I think

**Comment by cooljeanius** [comment]
> This code for making movement impossible might help:
> 
> ```
> #define MOVEMENT_RESTRICTION
>     # to avoid luring onto unprotected terrain, can still move in castles
>     [object]
>         silent=yes
>         duration=scenario
>         [effect]
>             apply_to=movement_costs
>             replace=true
>             [movement_costs]
>                 flat={UNREACHABLE}
>                 frozen={UNREACHABLE}
>                 cave={UNREACHABLE}
>                 reef={UNREACHABLE}
>                 sand={UNREACHABLE}
>                 shallow_water={UNREACHABLE}
>                 swamp_water={UNREACHABLE}
>                 deep_water={UNREACHABLE}
>             [/movement_costs]
>         [/effect]
>     [/object]
> #enddef
> ```
> 
> We could just get rid of the others and make 'frozen=1' or something, I think

Yeah that'd work for the movement part of it, but a pickuppable object requires more to it besides just the code for its effect... like, for snowshoes, I'd want to have an image for them, too...

**Comment by brirush84** [comment]
I messed around with some ideas for snowshoe images:

![snowshoes2](https://user-images.githubusercontent.com/28042598/127754467-4830024a-a1ec-41f0-96b9-ebd8e34fcb5d.png)

And I figured out how to code them and other pick-up-able items, if you're still interested in this avenue. It might be easier to do a 'ring of warmth' or something though.

**Comment by cooljeanius** [comment]
> I messed around with some ideas for snowshoe images:
> 
> ![snowshoes2](https://user-images.githubusercontent.com/28042598/127754467-4830024a-a1ec-41f0-96b9-ebd8e34fcb5d.png)
> 
> And I figured out how to code them and other pick-up-able items, if you're still interested in this avenue. It might be easier to do a 'ring of warmth' or something though.

ok I'll add adding this to my TODO list

**Comment by brirush84** [comment]
```
#define OBJ_SNOWSHOES X Y ID
    {PLACE_IMAGE items/snowshoes.png ({X}) ({Y})}
    [event]
        name=moveto
        first_time_only=no
        [filter]
            side=1
            x,y={X},{Y}
        [/filter]

 		[message]
			speaker=narrator
			image="wesnoth-icon.png"
			message= _ "Should $unit.name| pick up the snowshoes?"
			[option]
				label= _ "Yes"
				[command]
					[object]
						id={ID}
						name= _ "Snowshoes"
						image="icons/boots_elven.png"
						description= _ "A pair of snowshoes allowing any unit to walk on the snow!"
						cannot_use_message= _ "Only a powerful warrior may don this armor!"
						duration=forever
						[filter]
							x,y={X},{Y}
						[/filter]
						[then]
							[remove_item]
								x,y={X},{Y}
							[/remove_item]
							[message]
								speaker=$unit.id|
								image="wesnoth-icon.png"
								caption=""
								male_message= _ "$unit.name| pulls on the snowshoes. Now he can move freely on snow!"
								female_message= _ "$unit.name| pulls on the snowshoes. Now she can move freely on snow!"
							[/message]

							[modify_unit]
								[filter]
									x,y={X},{Y}
								[/filter]

								[trait]
									id=snow_shoes
									name= _ "snowshoes"
									description= _ "The snowshoes allow movement on frozen terrain."
								[/trait]
							[/modify_unit]
						[/then]
					  [effect]
						apply_to=movement_costs
						replace=true
						[movement_costs]
							frozen=1
						[/movement_costs]
					[/effect]
				 [/object]
				[/command]
			[/option]
			[option]
				label= _ "No"
			[/option]
		[/message]

    [/event]
 
#enddef
```

This should work (as long as the image path is alright). You should experiment with it; you can place multiple shoes with commands like:
```
		{OBJ_SNOWSHOES 22 21 shoes1}
		{OBJ_SNOWSHOES 22 32 shoes2}
		{OBJ_SNOWSHOES 22 31 shoes3}

```

**Comment by cooljeanius** [comment]
> 
> This should work (as long as the image path is alright). You should experiment with it

I did it slightly differently in 2d353e4; I wanted to use the `{PICKUPPABLE_ITEM}` macro. Should I add more snowshoes in the rest of the winter levels, or is just putting them in the first major winter level enough?

**Comment by cooljeanius** [comment]
Hm, one thing I've discovered while testing this, is that apparently frost stoats and icemonaxes don't get given unit names...

**Comment by brirush84** [comment]
You can fix that by saying unit.type instead of unit.name, I think

**Comment by cooljeanius** [comment]
> You can fix that by saying unit.type instead of unit.name, I think

But, the thing is, I *do* actually want to use the unit names when they're available though... is there a way I can do something like C's ternary operator? i.e., `($unit.name != "") ? $unit.name : $unit.type` or something...

Also, a separate issue I discovered is that units can pick up multiple pairs of snowshoes...

**Comment by cooljeanius** [comment]
I broke a transition in S14, "Back Home"... 
<img width="275" alt="Screen Shot 2021-08-09 at 01 42 47" src="https://user-images.githubusercontent.com/1388660/128664420-b2a8e837-a446-4c58-8924-3a01fad21262.png">

**Comment by cooljeanius** [comment]
> > You can fix that by saying unit.type instead of unit.name, I think
> 
> But, the thing is, I _do_ actually want to use the unit names when they're available though... is there a way I can do something like C's ternary operator? i.e., `($unit.name != "") ? $unit.name : $unit.type` or something...
> 
> Also, a separate issue I discovered is that units can pick up multiple pairs of snowshoes...

ok, the issue of being able to pick up multiple pairs of snowshoes is now fixed as of f3224db; the issue of icemonaxes and frost stoats not having names still remains, though... but anyways, the original issue of making mountains snowy is now fixed to my satisfaction, so I'll be closing this and opening a separate bug for the remaining issues.

**Comment by brirush84** [comment]
Okay, thanks! Sorry I dropped off, I started two online classes and have
stopped working on Wesnoth stuff. If I can work on the other stuff I will!
I’m glad you fixed that!

**Comment by cooljeanius** [comment]
> > But, the thing is, I _do_ actually want to use the unit names when they're available though... is there a way I can do something like C's ternary operator? i.e., `($unit.name != "") ? $unit.name : $unit.type` or something...
> > Also, a separate issue I discovered is that units can pick up multiple pairs of snowshoes...
> 
> ok, the issue of being able to pick up multiple pairs of snowshoes is now fixed as of [f3224db](https://github.com/cooljeanius/Son_Of_The_Black_Eye_Easy_Edition/commit/f3224db71054756ef80346683dd50a3a9d1b4cb8); the issue of icemonaxes and frost stoats not having names still remains, though... but anyways, the original issue of making mountains snowy is now fixed to my satisfaction, so I'll be closing this and opening a separate bug for the remaining issues.

The separate issue for the monster names is now issue #11.

> Okay, thanks! Sorry I dropped off, I started two online classes and have stopped working on Wesnoth stuff. If I can work on the other stuff I will! I’m glad you fixed that!

It's ok; thanks again for the icons! I added you to the credits in 2d353e4 btw. Do you think I should make another release now, now that I have this fixed? Or should I try to fix some more issues first?
(I noticed your version of the `{OBJ_SNOWSHOES}` macro added a trait to show which units had put them on, btw; while I left that out of my version, I'd accept PRs to get the trait to work with the `{PICKUPPABLE_ITEM}` macro, or to add some other way to display which units are wearing snowshoes; e.g. by adding your icon to the unit sprites)

**Comment by brirush84** [comment]
I would release new version if there are any other additional changes, but
if it's just the snowshoes maybe wait?

**Comment by cooljeanius** [comment]
> I would release new version if there are any other additional changes, but if it's just the snowshoes maybe wait?

Other changes, besides the snowshoes:

- added Dwarvish Woodcutters (from Default L0 Era)
- added Elvish Horsemen (from Default L0 Era)
- made mountains snowy (the original topic of this bug)
- used the "Over The Northern Mountains" music track in more scenarios
- Added a "kick" attack icon for use by Nightblades
- Gave more units more traits, for bug #8
- Made objectives for all allied sides
- Difficulty tweaks in some of the winter levels for bug #9 
- Fixed bug where having "The Great Horde" could cause level 0 units to give negative upkeep

**Comment by brirush84** [comment]
Yeah, I would make a new version then, sounds cool

**Comment by cooljeanius** [comment]
> Yeah, I would make a new version then, sounds cool

OK, I released 1.2.0: https://github.com/cooljeanius/Son_Of_The_Black_Eye_Easy_Edition/releases/tag/1.2.0
